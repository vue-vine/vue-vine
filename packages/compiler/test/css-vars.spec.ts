import { beforeEach, describe, expect, test } from 'vitest'
import hashId from 'hash-sum'
import type { VineCompilerHooks } from '../index'
import { compileVineTypeScriptFile, createCompilerCtx } from '../index'
import type { VineCompilerCtx } from '../src/types'

let mockVCF = ''
let mockCompilerHook = {} as VineCompilerHooks
let mockCompilerCtx = {} as VineCompilerCtx
beforeEach(() => {
  mockVCF = 'function testVCF() {\n'
    + '  const color = ref(\'red\')\n'
    + '\n'
    + '  vineStyle.scoped(`\n'
    + '    .blog-title {\n'
    + '      color: v-bind(color)\n'
    + '    }\n'
    + '  `)\n'
    + '  return vine`\n'
    + '     <div class="title">\n'
    + '        title\n'
    + '     </div>\n'
    + '  `\n'
    + '}'

  mockCompilerCtx = createCompilerCtx({})
  mockCompilerHook = {
    onOptionsResolved: cb => cb(mockCompilerCtx.options),
    onError: () => {},
    onWarn: () => {},
    onBindFileCtx: (fileId, fileCtx) => mockCompilerCtx.fileCtxMap.set(fileId, fileCtx),
  } as VineCompilerHooks
})
describe('CSS vars injection', () => {
  test('Should be injected based on reactive variables', () => {
    const res = compileVineTypeScriptFile(mockVCF, 'mockVCF', mockCompilerHook)
    const code = res.fileSourceCode.toString()
    expect(code.includes('useCssVars as _useCssVars')).toBeTruthy()
    expect(`'${hashId('testVCF' + 'color')}': (color.value)`).toBeTruthy()
    expect(code).toMatchSnapshot()
  })

  test('Should be injected based on normal variables', () => {
    const mockVCFNormal = mockVCF.replace('ref(\'red\')', '\'red\'')
    const res = compileVineTypeScriptFile(mockVCFNormal, 'mockVCFNormal', mockCompilerHook)
    const code = res.fileSourceCode.toString()
    expect(code.includes('useCssVars as _useCssVars')).toBeTruthy()
    expect(`'${hashId('testVCF' + 'color')}': (color)`).toBeTruthy()
    expect(code).toMatchSnapshot()
  })

  test('Should be injected based on props', () => {
    const mockVCFNormal = mockVCF.replace('ref(\'red\')', 'vineProp.withDefault(\'red\')')
    const res = compileVineTypeScriptFile(mockVCFNormal, 'mockVCFNormal', mockCompilerHook)
    const code = res.fileSourceCode.toString()
    expect(code.includes('useCssVars as _useCssVars')).toBeTruthy()
    expect(`'${hashId('testVCF' + 'color')}': (props.color)`).toBeTruthy()
    expect(code).toMatchSnapshot()
  })

  test('Should be able to inject according to each subcomponent of vfc', () => {
    const mockMultipleVCF = `${mockVCF}\n`
      + 'function testVCF2() {\n'
      + '  const color = ref(\'red\')\n'
      + '\n'
      + '  vineStyle.scoped(`\n'
      + '    .blog-title {\n'
      + '      color: v-bind(color)\n'
      + '    }\n'
      + '  `)\n'
      + '  return vine`\n'
      + '     <div class="title-s">\n'
      + '        title\n'
      + '     </div>\n'
      + '  `\n'
      + '}'

    const res = compileVineTypeScriptFile(mockMultipleVCF, 'mockVCFNormal', mockCompilerHook)
    const code = res.fileSourceCode.toString()
    expect(code.includes('useCssVars as _useCssVars')).toBeTruthy()
    expect(`'${hashId('testVCF' + 'color')}': (color.value)`).toBeTruthy()
    expect(`'${hashId('testVCF2' + 'color')}': (color.value)`).toBeTruthy()
    expect(code).toMatchSnapshot()
  })

  test('Should work with w/ complex expression', () => {
    const mockComplex = 'function testVCFComplex() {\n'
      + '  let a = 100\n'
      + '  let b = 200\n'
      + '  let foo = 300\n'
      + '\n'
      + '  vineStyle.scoped(`\n'
      + '     p {\n'
      + '       width: calc(v-bind(foo) - 3px);\n'
      + '       height: calc(v-bind(\'foo\') - 3px);\n'
      + '       top: calc(v-bind(foo + \'px\') - 3px);\n'
      + '     }\n'
      + '     div {\n'
      + '       color: v-bind((a + b) / 2 + \'px\' );\n'
      + '     }\n'
      + '     div {\n'
      + '       color: v-bind    ((a + b) / 2 + \'px\' );\n'
      + '     }\n'
      + '     p {\n'
      + '       color: v-bind(((a + b)) / (2 * a));\n'
      + '     }\n'
      + '  `)\n'
      + '  return vine`\n'
      + '     <div class="title">\n'
      + '        title\n'
      + '     </div>\n'
      + '  `\n'
      + '}\n'
    const res = compileVineTypeScriptFile(mockComplex, 'testVCFComplex', mockCompilerHook)
    const code = res.fileSourceCode.toString()
    expect(code.includes('useCssVars as _useCssVars')).toBeTruthy()
    expect(`'${hashId('testVCFComplex' + 'a')}': (a)`).toBeTruthy()
    expect(`'${hashId('testVCFComplex' + 'b')}': (b)`).toBeTruthy()
    expect(`'${hashId('testVCFComplex' + 'c')}': (c)`).toBeTruthy()
    expect(code).toMatchSnapshot()
  })
})
