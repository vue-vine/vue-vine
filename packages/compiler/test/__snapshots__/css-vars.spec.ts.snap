// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`CSS vars injection > Should be able to inject according to each subcomponent of vfc 1`] = `
"import {
  openBlock as _openBlock,
  createElementBlock as _createElementBlock,
  pushScopeId as _pushScopeId,
  popScopeId as _popScopeId,
  defineComponent as _defineComponent,
  useCssVars as _useCssVars
} from 'vue'
import 'mockVCFNormal?type=vine-style&scopeId=c9ba2a26&lang=css&scoped=true&virtual.css'
import 'mockVCFNormal?type=vine-style&scopeId=6d8b1a36&lang=css&scoped=true&virtual.css'


const testVCF = (() => {

const _withScopeId = n => (_pushScopeId(\\"data-v-c9ba2a26\\"),n=n(),_popScopeId(),n)
const _hoisted_1 = { class: \\"title\\" }

  const __vine = _defineComponent({
    name: 'testVCF',
    /* No props */
    /* No emits */
    setup(__props /* No setup ctx destructuring */) {

const props = __props


_useCssVars(_ctx => ({
  '2ad0dd52': (color.value),

}))

const color = ref('red')

/* No expose */

return (_ctx: any,_cache: any) => {
  return (_openBlock(), _createElementBlock(\\"div\\", _hoisted_1, \\" title \\"))
}

    }, // End of setup function
  }) // End of component object



  __vine.__scopeId = 'data-v-c9ba2a26'
  return __vine /* End of testVCF */
})()

const testVCF2 = (() => {

const _withScopeId = n => (_pushScopeId(\\"data-v-6d8b1a36\\"),n=n(),_popScopeId(),n)
const _hoisted_1 = { class: \\"title-s\\" }

  const __vine = _defineComponent({
    name: 'testVCF2',
    /* No props */
    /* No emits */
    setup(__props /* No setup ctx destructuring */) {

const props = __props


_useCssVars(_ctx => ({
  '49f4e33c': (color.value),

}))

const color = ref('red')

/* No expose */

return (_ctx: any,_cache: any) => {
  return (_openBlock(), _createElementBlock(\\"div\\", _hoisted_1, \\" title \\"))
}

    }, // End of setup function
  }) // End of component object



  __vine.__scopeId = 'data-v-6d8b1a36'
  return __vine /* End of testVCF2 */
})()"
`;

exports[`CSS vars injection > Should be injected based on normal variables 1`] = `
"import {
  openBlock as _openBlock,
  createElementBlock as _createElementBlock,
  pushScopeId as _pushScopeId,
  popScopeId as _popScopeId,
  defineComponent as _defineComponent,
  useCssVars as _useCssVars
} from 'vue'
import 'mockVCFNormal?type=vine-style&scopeId=c9ba2a26&lang=css&scoped=true&virtual.css'


const testVCF = (() => {

const _withScopeId = n => (_pushScopeId(\\"data-v-c9ba2a26\\"),n=n(),_popScopeId(),n)
const _hoisted_1 = { class: \\"title\\" }
const color = 'red'

  const __vine = _defineComponent({
    name: 'testVCF',
    /* No props */
    /* No emits */
    setup(__props /* No setup ctx destructuring */) {

const props = __props


_useCssVars(_ctx => ({
  '2ad0dd52': (color),

}))



/* No expose */

return (_ctx: any,_cache: any) => {
  return (_openBlock(), _createElementBlock(\\"div\\", _hoisted_1, \\" title \\"))
}

    }, // End of setup function
  }) // End of component object



  __vine.__scopeId = 'data-v-c9ba2a26'
  return __vine /* End of testVCF */
})()"
`;

exports[`CSS vars injection > Should be injected based on props 1`] = `
"import {
  openBlock as _openBlock,
  createElementBlock as _createElementBlock,
  pushScopeId as _pushScopeId,
  popScopeId as _popScopeId,
  defineComponent as _defineComponent,
  useCssVars as _useCssVars,
  toRefs as _toRefs
} from 'vue'
import 'mockVCFNormal?type=vine-style&scopeId=c9ba2a26&lang=css&scoped=true&virtual.css'

import { useDefaults as _useDefaults } from \\"vue-vine\\"

const testVCF = (() => {

const _withScopeId = n => (_pushScopeId(\\"data-v-c9ba2a26\\"),n=n(),_popScopeId(),n)
const _hoisted_1 = { class: \\"title\\" }

  const __vine = _defineComponent({
    name: 'testVCF',
    props: {
      color: { required: true },
    },
    /* No emits */
    setup(__props /* No setup ctx destructuring */) {

const props = _useDefaults(__props, {
  color: 'red'
})

const { color } = _toRefs(props)

_useCssVars(_ctx => ({
  '2ad0dd52': (props.color),

}))



/* No expose */

return (_ctx: any,_cache: any) => {
  return (_openBlock(), _createElementBlock(\\"div\\", _hoisted_1, \\" title \\"))
}

    }, // End of setup function
  }) // End of component object



  __vine.__scopeId = 'data-v-c9ba2a26'
  return __vine /* End of testVCF */
})()"
`;

exports[`CSS vars injection > Should be injected based on reactive variables 1`] = `
"import {
  openBlock as _openBlock,
  createElementBlock as _createElementBlock,
  pushScopeId as _pushScopeId,
  popScopeId as _popScopeId,
  defineComponent as _defineComponent,
  useCssVars as _useCssVars
} from 'vue'
import 'mockVCF?type=vine-style&scopeId=11bda7a6&lang=css&scoped=true&virtual.css'


const testVCF = (() => {

const _withScopeId = n => (_pushScopeId(\\"data-v-11bda7a6\\"),n=n(),_popScopeId(),n)
const _hoisted_1 = { class: \\"title\\" }

  const __vine = _defineComponent({
    name: 'testVCF',
    /* No props */
    /* No emits */
    setup(__props /* No setup ctx destructuring */) {

const props = __props


_useCssVars(_ctx => ({
  '2ad0dd52': (color.value),

}))

const color = ref('red')

/* No expose */

return (_ctx: any,_cache: any) => {
  return (_openBlock(), _createElementBlock(\\"div\\", _hoisted_1, \\" title \\"))
}

    }, // End of setup function
  }) // End of component object



  __vine.__scopeId = 'data-v-11bda7a6'
  return __vine /* End of testVCF */
})()"
`;

exports[`CSS vars injection > Should work with w/ complex expression 1`] = `
"import {
  openBlock as _openBlock,
  createElementBlock as _createElementBlock,
  pushScopeId as _pushScopeId,
  popScopeId as _popScopeId,
  defineComponent as _defineComponent,
  useCssVars as _useCssVars
} from 'vue'
import 'testVCFComplex?type=vine-style&scopeId=2b09f764&lang=css&scoped=true&virtual.css'


const testVCFComplex = (() => {

const _withScopeId = n => (_pushScopeId(\\"data-v-2b09f764\\"),n=n(),_popScopeId(),n)
const _hoisted_1 = { class: \\"title\\" }

  const __vine = _defineComponent({
    name: 'testVCFComplex',
    /* No props */
    /* No emits */
    setup(__props /* No setup ctx destructuring */) {

const props = __props


_useCssVars(_ctx => ({
  '5d7b94b2': (foo),

}))

let a = 100
let b = 200
let foo = 300

/* No expose */

return (_ctx: any,_cache: any) => {
  return (_openBlock(), _createElementBlock(\\"div\\", _hoisted_1, \\" title \\"))
}

    }, // End of setup function
  }) // End of component object



  __vine.__scopeId = 'data-v-2b09f764'
  return __vine /* End of testVCFComplex */
})()"
`;
